python/platforms/transition.bzl

```diff
@@ -0,0 +1,178 @@
+    # https://github.com/bazelbuild/bazel/blob/ce611646/src/main/java/com/google/devtools/build/lib/analysis/LocationExpander.java#L415-L429
+    #
+    # Since the default LocationExpander used by ctx.expand_location is not the same as the native
+    # rules (it doesn't set "allowDataAttributeEntriesInLabel"), we use "tools" temporarily while a
+    # proper fix in Bazel happens.
+    #
+    # A fix for this was proposed in https://github.com/bazelbuild/bazel/pull/16381.
+    "tools": attr.label_list(
+        allow_files = True,
+        mandatory = False,
+    ),
+    # Required to Opt-in to the transitions feature.
+    "_allowlist_function_transition": attr.label(
+        default = "@bazel_tools//tools/allowlists/function_transition_allowlist",
+    ),
+}
+
+_transition_py_binary = rule(
+    _transition_py_impl,
+    attrs = _COMMON_ATTRS,
+    executable = True,
+)
+
+_transition_py_test = rule(
+    _transition_py_impl,
+    attrs = _COMMON_ATTRS,
+    test = True,
+)
+
+def _py_rule(rule, transition_rule, name, target_platform, **kwargs):
+    args = kwargs.pop("args", None)
+    data = kwargs.pop("data", None)
+    env = kwargs.pop("env", None)
+
+    # Attributes common to all build rules.
+    # https://bazel.build/reference/be/common-definitions#common-attributes
+    compatible_with = kwargs.pop("compatible_with", None)
+    deprecation = kwargs.pop("deprecation", None)
+    distribs = kwargs.pop("distribs", None)
+    exec_compatible_with = kwargs.pop("exec_compatible_with", None)
+    exec_properties = kwargs.pop("exec_properties", None)
+    features = kwargs.pop("features", None)
+    restricted_to = kwargs.pop("restricted_to", None)
+    tags = kwargs.pop("tags", None)
+    target_compatible_with = kwargs.pop("target_compatible_with", None)
+    testonly = kwargs.pop("testonly", None)
+    toolchains = kwargs.pop("toolchains", None)
+    visibility = kwargs.pop("visibility", None)
+
+    rule(
+        name = "_" + name,
+        args = args,
+        data = data,
+        env = env,
+
+        # Attributes common to all build rules.
+        compatible_with = compatible_with,
+        deprecation = deprecation,
+        distribs = distribs,
+        exec_compatible_with = exec_compatible_with,
+        exec_properties = exec_properties,
+        features = features,
+        restricted_to = restricted_to,
+        tags = ["manual"] + (tags if tags else []),
+        target_compatible_with = target_compatible_with,
+        testonly = testonly,
+        toolchains = toolchains,
+        visibility = ["//visibility:private"],
+        **kwargs
+    )
+
+    return transition_rule(
+        name = name,
+        tools = data,
+        args = args,
+        env = env,
+        is_windows = select({
+            "@platforms//os:windows": True,
+            "//conditions:default": False,
+        }),
+        target = ":_" + name,
+
+        # Attributes common to all build rules.
```

Comment:
It seems like there's a lot of copy/paste for these common attributes. How about instead of manually carrying around ~11 variables, we stick them in a dict? 11 is a lot to carry around and makes it easy to forget one.

Reply:
Addressed in 161f068.